### PR TITLE
Use LittleEndian for correct byte order on DNS IPs

### PR DESCRIPTION
**Pull request by eduardok**
Nameserver IP 10.11.12.13 would otherwise get written to resolv.conf as 13.12.11.10, as was happening on my client.

Signed-off-by: Eduardo Kienetz <eduardo@kienetz.com>

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/568)

### CODE REVIEW

**Comment by bradfitz** [comment]
@dshynkev found https://bugzilla.redhat.com/show_bug.cgi?id=1153559 which suggests either endianness might not be the right answer and we might have to use the endianness of the host machine.

In other words, this change might break big endian machines. (And we do distribute mips binaries)

**Comment by dshynkev** [comment]
As written, [documentation](https://developer.gnome.org/NetworkManager/stable/settings-ipv4.html) seems to clearly suggest big endian ("network-byte-order"). I'd like to take some time to understand what's happening here, but I should get back soon.

**Comment by eduardok** [comment]
> @dshynkev found https://bugzilla.redhat.com/show_bug.cgi?id=1153559 which suggests either endianness might not be the right answer and we might have to use the endianness of the host machine.
> 
> In other words, this change might break big endian machines. (And we do distribute mips binaries)

Yes, I figured, feel free to close this pull request.

**Comment by eduardok** [comment]
Now that I think about it... since the DNS servers I'm getting are set on the portal, it means you are reading them and not storing as BigEndian, which is why when you go read ip.As16() you can't binary.BigEndian.Uint32() it. So this has nothing to do with Network Manager, and this PR should be dropped.

**Comment by eduardok** [comment]
That is, DNSConfig.Nameservers needs fixing.
I wonder if wgIPToNetaddr() should be using netaddr.FromStdIPRaw() instead of netaddr.FromStdIP(), assuming Raw is properly as BigEndian. Anyway, too late for me to think about this, good luck.

**Comment by dshynkev** [comment]
It seems that the documentation is simply lying: the DBus API specification states that the endianness of our messages should match the endianness of the messages from the other side (NetworkManager), which is exposed as a flag in the first byte of the message (see [DecodeMessage](https://github.com/godbus/dbus/blob/master/message.go#L136)).

Unfortunately, we cannot easily get at that flag because `godbus` doesn't expose it and doesn't let us look at raw messages. I'll file an issue with them, but in practice it should just be host endianness.

In other words, this PR is doing essentially the right thing, it just needs to use native endian, e.g. like [this](https://github.com/godbus/dbus/blob/master/transport_generic.go#L10).

**Comment by dshynkev** [comment]
It seems that the documentation is simply lying: the DBus API specification states that the endianness of our messages should match the endianness of the messages from the other side (NetworkManager), which is exposed as a flag in the first byte of the message (see [DecodeMessage](https://github.com/godbus/dbus/blob/master/message.go#L136)).

Unfortunately, we cannot easily get at that flag because `godbus` doesn't expose it and doesn't let us look at raw messages. I'll file an issue with them, but in practice it should just be host endianness.

In other words, this PR is doing essentially the right thing, it just needs to use native endian, e.g. like [this](https://github.com/godbus/dbus/blob/master/transport_generic.go#L10).

I don't think there is any problem with `netaddr.IP` values themselves, as that would come through everywhere else too (e.g. `IP.String()` is `fmt.Sprintf("%d.%d.%d.%d", ip[0], ip[1], ip[2], ip[3])`, and that is correct).

**Comment by eduardok** [comment]
I haven't looked deep enough to find out how you get the DNS servers we set on the management portal (https://login.tailscale.com/admin/dns), but I wouldn't think you are storing in a NetworkManager data structure. Yet, even if you were, it looks like if you stored as BigEndian, it would read correctly...

**Comment by dshynkev** [comment]
We are not storing them in a NetworkManager data structure at any point until this one. In the network map we get from control, which is JSON-encoded, they are marshaled as strings. I'm pretty sure we don't convert them to `uint32` anywhere else along the way either (they can be IPv6 addresses!), and the other dns modes work (e.g. my device uses resolvconf with no issues), so I think it's fair to say that the issue must be localized here.

**Comment by eduardok** [comment]
Ok, I guess you are right in that netaddr is fine, it's just that go's net.IP doesn't consider endianess (at least not in the context we are seeing here):
func IPv4(a, b, c, d byte) IP {
        p := make(IP, IPv6len)
        copy(p, v4InV6Prefix)
        p[12] = a
        p[13] = b
        p[14] = c
        p[15] = d
        return p
If I'm looking at the right places...
So, I guess you guys just need to decide if you will store as big endian or not, considering other platforms using NetworkManager, then my PR might apply dependent on that.

**Comment by eduardok** [comment]
> I'm fine with merging this as-is, too. This is a definite improvement over breaking approximately everyone using NetworkManager. We can solve the MIPS issue by properly reading the message flags, which will take some time to get solved on the godbus end.
> 
> In the meantime, we can do this and avoid the unsafe endian-detecting hack.

Makes sense to me, thanks.

**Comment by danderson** [comment]
Yup, netaddr and the standard library net.IP both store stuff in big-endian form. We carry IP addrs around the codebase either in string form, or in one of those two containers that has a consistent byte encoding for when we need it. It only gets massaged into other forms when we have to interact with the various whacky APIs the various OSes give us.

This change as-is LGTM. Doing the native-endian thing is the 100% fix, but this takes us from 99% of tailscale+networkmanager users being broken to 99% working. We can catch the last percent in a followup, once godbus gives us the info we need.

Thanks Eduardo!
